pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -508,6 +560,77 @@ pub mod pallet {
 			})
 		}
 
+		// TODO: this weight should likely scale based on collect_n_epochs
+		#[pallet::weight(100)]
+		pub fn collect(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			tranche_id: T::TrancheId,
+			collect_n_epochs: u32,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let loc = TrancheLocator {
+				pool_id,
+				tranche_id,
+			};
+			let order = Order::<T>::try_get(&loc, &who).unwrap_or(UserOrder {
+				supply: Zero::zero(),
+				redeem: Zero::zero(),
+				epoch: Zero::zero(),
+			});
+			let pool = Pool::<T>::try_get(pool_id)
+				.map_err(|_| Error::<T>::NoSuchPool)
+				.unwrap();
+			ensure!(
+				order.epoch <= pool.last_epoch_executed,
+				Error::<T>::EpochNotExecutedYet
+			);
+
+			let n_epochs: T::EpochId = collect_n_epochs.into();
+			let end_epoch: T::EpochId = if order.epoch + n_epochs > pool.last_epoch_executed {
+				pool.last_epoch_executed
+			} else {
+				(order.epoch + n_epochs).into()
+			};
+
+			let collections = Self::calculate_collect(loc.clone(), order, pool, end_epoch);
+			let pool_account = PoolLocator { pool_id }.into_account();
+
+			if collections.payout_currency_amount > Zero::zero() {
+				let pool = Pool::<T>::try_get(pool_id).map_err(|_| Error::<T>::NoSuchPool)?;
+				T::Tokens::transfer(
+					pool.currency,
+					&pool_account,
+					&who,
+					collections.payout_currency_amount,
+				)?;
+			}
+
+			if collections.payout_token_amount > Zero::zero() {
+				let token = T::TrancheToken::tranche_token(pool_id, tranche_id);
+				T::Tokens::transfer(token, &pool_account, &who, collections.payout_token_amount)?;
+			}
+
+			Order::<T>::try_mutate(&loc, &who, |order| -> DispatchResult {
+				order.supply = collections.remaining_supply_currency;
+				order.redeem = collections.remaining_redeem_token;
+				order.epoch = end_epoch + One::one();
+
+				Self::deposit_event(Event::OrdersCollected(
+					pool_id,
+					tranche_id,
+					end_epoch,
+					who.clone(),
+					collections.payout_currency_amount,
+					collections.payout_token_amount,
+					collections.remaining_supply_currency,
+					collections.remaining_redeem_token,
+				));
```

Comment:
Move this out of the `try_mutate` maybe?

Reply:
I'm not sure if there are any considerations with this besides code cleanliness, but I put it here because I noticed the `deposit_event` is put within `try_mutate` in several other places in the code already.